print detailed error info for docker pull

WeiZhang555 opened the pull request:
When docker push get response with unknown HTTP status, docker daemon print:
"Error: Status XXX trying to push repository XXX: XXX"
But when docker pull meets response with unknown status code, it gives:
"HTTP code: XXX"

This commit helps docker pull print more detailed error info like push
does, so push and pull can behave consistently when error happens.

We are developing our own private hub, this print is especially helpful because we can
define our own status codes for different error and give users more infos of what's wrong
with their operation. Hope that you can understand and add this in

BTW, I'm not sure how to write the test case, does this also need a test case? 

Signed-off-by: Zhang Wei zhangwei555@huawei.com

dmcgowan commented:
This change shouldn't require an additional test case since it is very hard to replicate in the current integration tests.

dmcgowan commented:
What does this show on the client side for you? I find this will ensure something useful is printed in the daemon logs, but the client will only end up seeing `FATA[0000]` with nothing after it when res is multi-lined.

HammerOfThor commented:
@dmcgowan for example, if user run "docker pull 10.110.11.123/jack/busybox,  server 10.110.11.123 respond with HTTP status code 405 and HTTP body "you access wrong namespace", then docker client will print "FATA[0000] Error: Status 405 trying to pull repository jack/busybox: you access wrong namespace", then user will know he visit a wrong namespace/repo. Docker daemon log will print same message.

dmcgowan commented:
The problem is if the registry responds with html content or something with newlines, only the last line will be visible cli, which is usually empty. Giving you just "FATA[0000]". This is not a problem with your PR since it occurs with many other endpoints, just made worse in that case since the HTTP code will no longer be seen.  Even in that case this PR is still an improvement though since the daemon side logs will show the response body, making it easier to debug. I think there is a general problem with the cli that is not properly showing the logs in this case, need to look further into it.

HammerOfThor commented:
@dmcgowan 
Yes, you are right, I reproduce this senario successfully, if http respond with a newline character at the end of body, docker client will print nothing.
Not just docker pull has this issue, docker push has same problem, so this is a generalized problem.
Maybe we can fix this issue after this PR? 
What's your opinion?  :)

dmcgowan commented:
I agree, we can fix all of these in bulk in a separate PR.  It might be as easy as using `%q` when formatting the content of the body. Can you make a PR or issue for this and ping me?  If you can't let me know and I can get to it later in the week.  Thanks!

LGTM

HammerOfThor commented:
@dmcgowan OK, I'll send an issue and ping you :)

dmcgowan commented:
ping @jlhawn @stevvooe for second review

stevvooe commented on registry/session.go:
```diff
@@ -281,7 +281,11 @@ func (r *Session) GetRepositoryData(remote string) (*RepositoryData, error) {
-		return nil, utils.NewHTTPRequestError(fmt.Sprintf("HTTP code: %d", res.StatusCode), res)
+		errBody, err := ioutil.ReadAll(res.Body)
+		if err != nil {
+			return nil, err
```
This will hide the original application error is there is a problem reading the body. This should still return a `utils.NewHTTPRequestError` error, but perhaps without the body.

There are lots of cases where you get a non-200 response with an empty body. This code will fail to report that in a meaningful way to the user.

HammerOfThor replied:
@stevvooe Yeah, I got what you say, making a second PR....

HammerOfThor commented:
@stevvooe I make some improvement, how about this one?
Actually my code refers to "docker push" http error response handler, e.g. in function PushImageJSONIndex():
514         if res.StatusCode != 200 && res.StatusCode != 201 {
515             errBody, err := ioutil.ReadAll(res.Body)
516             if err != nil {
517                 return nil, err
518             }
519             return nil, utils.NewHTTPRequestError(fmt.Sprintf("Error: Status %d trying to push repository %s: %s", res.StatusCode, remote, errBody    ), res)
520         }

Maybe I can make another PR to change this as you say, what do you think?

HammerOfThor commented:
Another question, my PR seems to build failed on windows, does this matter? Seems not my problem...

stevvooe commented on registry/session.go:
```diff
@@ -281,7 +281,11 @@ func (r *Session) GetRepositoryData(remote string) (*RepositoryData, error) {
 	if res.StatusCode != 200 {
-		return nil, utils.NewHTTPRequestError(fmt.Sprintf("HTTP code: %d", res.StatusCode), res)
+		errBody, err := ioutil.ReadAll(res.Body)
+		if err != nil || string(errBody) == "" {
```
Use `len(errBody) == 0` so the whole buffer isn't copied.

stevvooe commented:
@HammerOfThor That code has the same issue. If they're related, put them in the same PR.

HammerOfThor commented:
@stevvooe Another version, let me know what you feel about this :) Thank you!

dmcgowan commented on registry/session.go:
```diff
@@ -281,7 +281,11 @@ func (r *Session) GetRepositoryData(remote string) (*RepositoryData, error) {
 	if res.StatusCode != 200 {
-		return nil, utils.NewHTTPRequestError(fmt.Sprintf("HTTP code: %d", res.StatusCode), res)
+		errBody, err := ioutil.ReadAll(res.Body)
+		if err != nil || len(errBody) == 0 {
```
The error is essentially being ignored in this case. If the err is only informational for not used by the caller, then it should just be logged and only one return should occur.

Essentially
~~~
if err != nil  {
	log.Debugf("Error reading response body: %s", err)
}
return nil, utils.NewHTTPRequestError(fmt.Sprintf("Error: Status %d trying to pull repository %s: %s", res.StatusCode, remote, errBody), res)
~~~

stevvooe replied:
@HammerOfThor Sorry for the mixed messaging, but I have to agree with @dmcgowan. I hope it is not too much trouble.

WeiZhang555 committed:
print detailed error info for docker pull

When docker push get response with unknown HTTP status, docker daemon
print:
"Error: Status XXX trying to push repository XXX: XXX"
But when docker pull meets response with unknown status code, it gives:
"HTTP code: XXX"

This commit helps docker pull print more detailed error info like push
does, so push and pull can behave consistently when error happens.

Signed-off-by: Zhang Wei <zhangwei555@huawei.com>

HammerOfThor commented:
never mind, hope that I make it right now. 
ping @dmcgowan @stevvooe

stevvooe commented:
LGTM on the code.

@jfrazelle I am not sure what is going on with the failure. It doesn't look related but may be if timing out on reading the body.

dmcgowan commented:
LGTM

jlhawn commented:
LGTM

dmcgowan commented on registry/session.go:
```diff
@@ -281,7 +281,11 @@ func (r *Session) GetRepositoryData(remote string) (*RepositoryData, error) {
+		if err != nil {
+			log.Debugf("Error reading response body: %s", err)
+		}
+		return nil, utils.NewHTTPRequestError(fmt.Sprintf("Error: Status %d trying to pull repository %s: %s", res.StatusCode, remote, errBody), res)
```
Updated the other calls to use `%q` here for errBody, this protects against the new line issue. Can you make this update and we can get it merged

stevvooe replied:
`errBody` would have to be cast to a string for that to work correctly. Right now, only a raw byte array would be dumped.

HammerOfThor replied:
@dmcgowan update done
@stevvooe errBody should be meaningful info returned by hub, I think ```%q``` can work fine after some simple tests. :)

stevvooe replied:
@HammerOfThor SGTM!

dmcgowan replied:
Agreed in my testing `.String()` was not needed.

dmcgowan commented:
Thank you for updating, you will need to rebase as well, since the `%q` change is in master it should be a clean resolve

stevvooe commented:
LGTM!

HammerOfThor commented:
@dmcgowan rebase done, now code is clean

dmcgowan commented:
LGTM

duglin commented:
LGTM

HammerOfThor commented:
Thank you all for helping reviewing my code
@dmcgowan @stevvooe  and other guys  :)
